### PR TITLE
gromacs: add cufftmp variant to enable distributed PME when using CUDA

### PR DESCRIPTION
**Pull request by danielahlin**

### CODE REVIEW

**Review by junghans — DISMISSED** [review]

**Review by marvinbernhardt — DISMISSED** [review]
Not tested, but looks right.

**Review by alalazo — DISMISSED** [review]

**Comment by w8jcik** [comment]
I think cufftmp is only used with MPI enabled. If I am correct, maybe a constraint should be added.

**Comment by alalazo** [comment]
@danielahlin Can you check https://github.com/spack/spack/pull/36552#issuecomment-1495922987 ?

**Comment by danielahlin** [comment]
>> I think cufftmp is only used with MPI enabled. If I am correct, maybe a constraint should be added.

> @danielahlin Can you check [#36552 (comment)](https://github.com/spack/spack/pull/36552#issuecomment-1495922987) ?

@w8jcik to make sure I'm on the right track before testing -  are you thinking of something like:

variant("cufftmp", default=False, description="Enable Multi GPU FFT support", when="+cuda+mpi")

or

conflicts( "~cuda", when="+cufftmp", description="Cuda FFT MP can only be used with cuda enabled")
conflicts( "~mpi", when="+cufftmp", description="Cuda FFT MP can only be used with mpi enabled")

or 

depends_on("mpi", when="+cufftmp")
depends_on("cuda", when="+cufftmp")


(I added cuda since it is also required for this variant.)

**Comment by w8jcik** [comment]
I think the least intrusive is to go with conflicts

```python
conflicts( "~cuda", when="+cufftmp", description="Cuda FFT MP can only be used with cuda enabled")
conflicts( "~mpi", when="+cufftmp", description="Cuda FFT MP can only be used with mpi enabled")
```

But other also fine with me.

**Comment by w8jcik** [comment]
After installation, you should see

```
gmx_mpi --version

...
MPI library:        MPI (GPU-aware: CUDA)
GPU support:        CUDA
GPU FFT library:    cuFFT
Multi-GPU FFT:      cuFFTMp
```

If I remember correctly it builds successfully without MPI, but cuFFTMp is not used and it won't be here. Not 100% sure.

File: var/spack/repos/builtin/packages/gromacs/package.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ class Gromacs(CMakePackage, CudaPackage):
         description="Produces a double precision version of the executables",
     )
     variant("plumed", default=False, description="Enable PLUMED support")
+    variant("cufftmp", default=False, description="Enable Multi GPU FFT support")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gromacs/package.py]: If the semantic is the one discussed in the comments, then:
```suggestion
    variant("cufftmp", default=False, when="+cuda+mpi", description="Enable Multi GPU FFT support")
```
This means that the variant exists only when `+cuda+mpi` are both active.

Comment (w8jcik) [reply var/spack/repos/builtin/packages/gromacs/package.py]: But do you mean that it is good this way, or bad?

Comment (alalazo) [reply var/spack/repos/builtin/packages/gromacs/package.py]: Whenever possible it's better to have conditional variants than conflicts (especially conflicts with default values) due to how the concretizer optimizes to reach the "best" possible solution.

**Comment by danielahlin** [comment]
Ok - I've made cufftmp conditional on 

* cuda - no doubt there
* mpi - less obvious since cufftmp should only depend on nvshmem - however since IIUC cufftmp is only necessary for multi-GPU FFTs when having a multi-process job and Gromacs only do multi-process with MPI it should be OK.

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@danielahlin Please read here https://github.com/spack/spack-infrastructure/blob/main/docs/deferred_pipelines.md (it's linked into the "Details", but not very visible unfortunately). The bottomline is that, the more you rebase / merge the more you have to potentially wait for pipelines :slightly_frowning_face:

**Comment by danielahlin** [comment]
@alalazo thanks - I'll keep this in mind! (I actually thought the process was blocked due to not being in sync.)

**Comment by alalazo** [comment]
> I actually thought the process was blocked due to not being in sync

Yeah, wrote the comment since I think having a "stale" yellow dot gives that impression. I wonder if we might have @spackbot posting a message to let users know how Gitlab pipelines are handled.

**Pull request merged**